src/server.c

```diff
@@ -994,3 +997,831 @@ dqlite_node_id dqlite_generate_node_id(const char *address)
 
 	return raft_digest(address, n);
 }
+
+/* Dynamic array of node info objects. This is the in-memory representation of
+ * the node store. */
+struct node_store_cache
+{
+	struct client_node_info *nodes; /* owned */
+	unsigned len;
+	unsigned cap;
+};
+
+struct dqlite_server
+{
+	/* Threading stuff: */
+	pthread_cond_t cond;
+	pthread_mutex_t mutex;
+	pthread_t refresh_thread;
+
+	/* This field is only accessed on the "main thread": */
+	bool started;
+
+	/* These fields are protected by the mutex: */
+	bool shutdown;
+	struct node_store_cache cache;
+	/* We try to keep this pointing at the leader, but it might be out of
+	 * date or not open. */
+	struct client_proto proto;
+
+	/* These fields are immutable after construction: */
+	bool is_new;
+	bool bootstrap;
+	char *dir_path; /* owned */
+	dqlite_node *local;
+	uint64_t local_id;
+	char *local_addr; /* owned */
+	char *bind_addr;  /* owned */
+	dqlite_connect_func connect;
+	void *connect_arg;
+	unsigned refresh_period;
```

Comment:
Also, I'd suggest to make this field store the value in milliseconds instead of seconds. This should be handy for integration tests, so we can have this logic trigger faster.